libsolidity/codegen/ir/IRGeneratorForStatements.cpp

```diff
@@ -1733,7 +1751,14 @@ void IRGeneratorForStatements::endVisit(Identifier const& _identifier)
 		return;
 	}
 	else if (FunctionDefinition const* functionDef = dynamic_cast<FunctionDefinition const*>(declaration))
-		define(_identifier) << to_string(functionDef->resolveVirtual(m_context.mostDerivedContract()).id()) << "\n";
+	{
+		FunctionDefinition const& resolvedFunctionDef = functionDef->resolveVirtual(m_context.mostDerivedContract());
+		define(_identifier) << to_string(resolvedFunctionDef.id()) << "\n";
+
+		solAssert(resolvedFunctionDef.functionType(true), "");
+		solAssert(resolvedFunctionDef.functionType(true)->kind() == FunctionType::Kind::Internal, "");
+		m_context.internalFunctionAccessed(_identifier);
```

Comment:
Why is it not a problem that you mark the function accessed via the expression (and `internalFunctionAccessed` then using `.referencedDeclaration`) but store the value after doing virtual function lookup?

Reply:
Because `internalFunctionAccessed()` gets `_identifer` (and extracts the function definition on its own), rather than `resolvedFunctionDef`. I'm using `resolvedFunctionDef` here only to check assertions.

I think I was using `resolvedFunctionDef` (possibly incorrectly) in one of the previous versions here so it made more sense back then but yeah, now it's misleading. I'll change the assertions to use `functionDef` instead.

Reply:
Done. Assertions are now using `functionDef`.

Reply:
I'm still confused. You assign the correct function id to the variable, so it will call the right function, but inside `internalFunctionAccessed` ( https://github.com/ethereum/solidity/pull/8797/files#diff-f841efc1479453da2e05b159a7f1da6bR165 ) you store the base function in the dispatch map and request code generation for the base function.

Shouldn't this go wrong in the following case:
```
contract A {
  function f() virtual internal {}
  function g() internal { f(); }
}
contract B is A {
  function f() override internal {} // code for this is never generated
  function h() public { g(); }
}
```

Reply:
Edited the above to force call via pointer.

Reply:
You're right it does not work with your example. I fixed it by rewriting `internalFunctionAccessed()` to accept a function definiton as well. That was how it was supposed to work back when I was still adding and removing candidates. I got this confused during one of the rewrites when I decided that I could get by just with the expression.

I added an example similar to yours as a test (`internal_virtual_function_calls_through_dispatch.sol`). Not exactly the same since the `(f)()` syntax makes an unrelated assertion fail, as I explained on gitter. This assertion is already in develop, so I'll be fixing it separately.

Anyway, this is another bug in this code that slipped through and I think I just need to start adding more tests. There are lots of tests already and I assumed that they cover most of the corner cases so I was adding them sparingly. But looks like I need a better coverage. It's hard to spot these errors myself when rewriting because of the assumptions I have from the previous versions...